R/cql-geom-predicates.R

```diff
@@ -128,6 +118,58 @@ sf_text <- function(x) {
   sf::st_as_text(x)
 }
 
+
+#' Check spatial objects for WFS spatial operations
+#'
+#' Check a spatial object to see if it exceeds the current set value of
+#' 'bcdata.max_geom_pred_size' option. If the object does exceed the size
+#' threshold a bounding box was drawn around the polygon and all features
+#' within the box will be returned. Further options include:
+#' - Try adjusting the value of the 'bcdata.max_geom_pred_size' option
+#' - Simplify the spatial object to reduce its size
+#' - Further processing on the returned object
+#'
+#' @param x object of class sf, sfc or sfg
+#'
+#' @return a logical indicating whether the option value was exceeded
+#' @export
+#'
+#' @examples
+#' \donttest{
+#' try({
+#'   airports <- bcdc_query_geodata("bc-airports") %>% collect()
+#'   bcdc_size_check(airports)
+#' })
+#' }
+bcdc_size_check <- function(x) {
+  if (!inherits(x, c("sf", "sfc", "sfg", "bbox"))) {
+    stop(paste(deparse(substitute(x)), "is not a valid sf object"),
+         call. = FALSE)
+  }
+
+  if (inherits(x, "bbox")) {
+    obj_size <- utils::object.size(x)
+  } else {
+    obj_size <- utils::object.size(sf::st_geometry(x))
+  }
+
+  option_size <- getOption("bcdata.max_geom_pred_size", 5E5)
+
+  ## If too big here, drawing bounding
+  if (obj_size > option_size) {
+
+    message(bold_blue("The object is too large to perform exact spatial operations using bcdata."))
+    message(bold_blue(glue::glue("Object size: {obj_size} bytes")))
```

Comment:
For the messages, I wonder if we should put them in MB (bytes * 1024 * 1024) for readability?

Reply:
I struggled with this and actually wondered if using the units package might be a good idea. We already have it as a dependency via sf.

Reply:
I think it might be overkill for this

Reply:
Now I am hesitant to change this unless we change the default value. For 5E5 is the default in _bytes_. So either we should change the units of the default value to MB _or_ leave the message in bytes. I think this might introduce so overhead that isn't worth it here. Happy to proven wrong though.

Reply:
Yeah it's probably fine as is. Pretty niche already!